### ISSUE TITLE
Programmatic API

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#935]
ESLint should have a programmatic API that let's you easily use it a normal node module. It should be fully decoupled from the CLI, and the CLI should use the programmatic API underneath.

Having to use the CLI interface directly comes with a lot of assumptions and stdout noise.

I'm not really sure what the API should look like as I'm not aware of all the use-cases, but it should be simple.

Maybe something like this:

``` js
var linter = new ESLint({
    files: [],
    formatter: require('eslint-json')
});

linter.on('warning', function (warning) {
    console.log('ESLint:', warning);
});

linter.run(function (err, result) {
    console.log(result);
    //=> { errors: { ... }, ... }
});
```

Nothing should be automatically outputted to stdout/stderr when using the programmatic API.

It should allow supplying both paths and file contents as strings, so to support all possible use-cases.

Having a real programmatic API is important to make it easy to integrate ESLint into existing systems.
